### PR TITLE
Fix win_write to check if fully sent

### PR DESCRIPTION
**Pull request by qbx2**
`win_write` does not check if the data is completely sent in synchronous phase. When that situation happens, the connection hangs up. I added handling code for that.

@donnadionne

### CODE REVIEW

**Comment by qbx2** [comment]
@donnadionne

**Comment by qbx2** [comment]
Please review.

**Comment by Schramp** [comment]
Fixes issue #28524

**Comment by Schramp** [comment]
I still want this. Any way to upvote this?

**Review by drfloob — COMMENTED** [review]
Thanks for this, sorry it slipped through the cracks (significantly). It looks correct to me, but I'm not sure how our tests never caught it. In particular, we have an end2end test that sends a 10MB payload (`//test/core/end2end:h2_full_test@invoke_large_request`). Are the windows transport buffers > 10MB by default? Best I could find, the default is O(kilobytes). If we can add a test for this, or modify one to catch it, that would be ideal.

File: src/core/lib/iomgr/tcp_windows.cc [hunk 1]
```diff
@@ -391,26 +391,48 @@ static void win_write(grpc_endpoint* ep, grpc_slice_buffer* slices,
...
 
   TCP_REF(tcp, "write");
 
   /* If we got a WSAEWOULDBLOCK earlier, then we need to re-do the same
      operation, this time asynchronously. */
   memset(&socket->write_info.overlapped, 0, sizeof(OVERLAPPED));
-  status = WSASend(socket->socket, buffers, (DWORD)tcp->write_slices->count,
+  status = WSASend(socket->socket, buffers + async_buffers_offset,
+                   (DWORD)(tcp->write_slices->count - async_buffers_offset),
                    &bytes_sent, 0, &socket->write_info.overlapped, NULL);
```
[the remark this answers is not in the document]
Comment (qbx2) [reply src/core/lib/iomgr/tcp_windows.cc]: Done

**Comment by qbx2** [comment]
Actually, the issue does not cause on windows environment because windows does not perform partial send (I didn’t know this at the time of writing code). It is only caused when we run grpc server on WINE. Even though, I think it’d be best to fix it because the behavior is not documented by microsoft. Therefore, we could use WINE to catch it. Do you think it is ok to add WINE tests to catch it?

**Comment by Schramp** [comment]
> It is only caused when we run grpc server on WINE. 
Same way I found it.

**Review by Schramp — APPROVED** [review]

File: src/core/lib/iomgr/tcp_windows.cc [hunk 2]
```diff
@@ -391,27 +391,49 @@ static void win_write(grpc_endpoint* ep, grpc_slice_buffer* slices,
...
 
   /* If we got a WSAEWOULDBLOCK earlier, then we need to re-do the same
      operation, this time asynchronously. */
   memset(&socket->write_info.overlapped, 0, sizeof(OVERLAPPED));
-  status = WSASend(socket->socket, buffers, (DWORD)tcp->write_slices->count,
-                   &bytes_sent, 0, &socket->write_info.overlapped, NULL);
+  status = WSASend(socket->socket, buffers + async_buffers_offset,
+                   (DWORD)(tcp->write_slices->count - async_buffers_offset),
+                   NULL, 0, &socket->write_info.overlapped, NULL);
   if (allocated) gpr_free(allocated);
```
Comment (Schramp) [inline src/core/lib/iomgr/tcp_windows.cc:437]: Please note that the blocking behaviour of the first WSASend function on Windows hides the use-after free of the variable "allocated" (or the stack allocated version "local_buffers" that gets out of scope) of the non blocking second WSASend. In Windows this code is never reached.

This is however another issue that is out of scope of this PR.

Comment (drfloob) [reply src/core/lib/iomgr/tcp_windows.cc:437]: I'm not sure that's right. Per the [WSASend docs](https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsasend):

> If this function is completed in an overlapped manner, it is the Winsock service provider's responsibility to capture the [WSABUF](https://learn.microsoft.com/en-us/windows/desktop/api/ws2def/ns-ws2def-wsabuf) structures before returning from this call. This enables applications to build stack-based WSABUF arrays pointed to by the lpBuffers parameter.

Deallocating the heap (and local) variables before returning shouldn't be an issue for the subsequent overlapped WSASend call. Do I maybe misunderstand your concern here?

**Comment by drfloob** [comment]
@qbx2, regarding:

> Actually, the issue does not cause on windows environment because windows does not perform partial send

 Do you have a source for that information? I'm not sure it's true, if the implementations are built to WSASend's current spec:

> For non-overlapped sockets, ... If the socket is non-blocking and stream-oriented, and there is not sufficient space in the transport's buffer, WSASend will return with only part of the application's buffers having been consumed.

**Comment by drfloob** [comment]
@Schramp @qbx2 Given that our concern about use-after-free can be considered separately, I'm fine to merge this PR as is. @veblush can you take a second look as well, please?

**Review by drfloob — APPROVED** [review]

**Review by veblush — APPROVED** [review]

**Comment by qbx2** [comment]
@drfloob The source is here: https://bugs.winehq.org/show_bug.cgi?id=10648 . Also, some WINE contributor said that windows implementation sometimes does not follow the specification (documentation)

File: src/core/lib/iomgr/tcp_windows.cc (hunk 2 above)
Comment (Schramp) [reply src/core/lib/iomgr/tcp_windows.cc:437]: You are right, the copying of the buffer in wsasend is new to me. I more or less assumed it was up to the completion handler.

My comments on use after free can be ignored.

**Pull request merged**